### PR TITLE
Vault EREC4626 implementation 

### PR DESCRIPTION
**Pull request by yotamdaniel**
Whats in this PR:

- hard copy of open zeppelin implementation IERC and ERC 4626. Will be removed once we can import it
- Vault OUSD contract
- Tests
- Reseting network before integration tests so all tests pass

Whats not in this PR:

- full suit of tests to test ERC4626

FOB @jasonstockman

### CODE REVIEW

File: contracts/VaultOUSD.sol [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+pragma solidity 0.8.4;
+
+import "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import {SafeERC20} from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+import {IERC4626} from "../contracts/interfaces/IERC4626.sol";
+import {ERC4626} from "../contracts/standard/ERC4626.sol";
+
+contract VaultOUSD is ERC4626 {
```
Comment (tomerXarchimedes) [inline contracts/VaultOUSD.sol:11]: I would add a comment around what this Vault is doing (like: only hold OUSD and using 4626 because it deals with re-basing)

File: test/VaultOUSD.js [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+    let tokenOUSD;
+    let owner;
+    let addr1;
+    let addr2;
+    let addrs;
+    let sharesOwnerAddress;
+
+    let addr1Deposit = 10000
+    let addr2Deposit = 20000
+    let interestIntoVault = 10000
```
Comment (tomerXarchimedes) [inline test/VaultOUSD.js:25]: OUSD decimal is 18 - are these numbers really small (like: 10,000 / 10^18) or these are 10K OUSD?
I would add a comment clarifying that (and if needed add the conversion to the correct decimal).
Otherwise we might forget what we meant next time we visit this test

File: test/VaultOUSD.js [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+            = await contractVault.deploy(tokenOUSD.address, "VaultOUSD", "VOUSD");
+
+        // Mint initial amount on OUSD token, will be used by all tests
+        await MainnetHelper.helperSwapETHWithOUSD(addr1, ethers.utils.parseEther("100.0"))
+        await MainnetHelper.helperSwapETHWithOUSD(addr2, ethers.utils.parseEther("200.0"))
+        await MainnetHelper.helperSwapETHWithOUSD(owner, ethers.utils.parseEther("300.0"))
+        sharesOwnerAddress = owner.address
+        // deposit OUSD as a user (that gets shares) into vault. Shares goes to owner, not user.
+        await tokenOUSD.connect(addr1).approve(tokenVault.address, addr1Deposit);
+        await tokenVault.connect(addr1).deposit(addr1Deposit, sharesOwnerAddress)
```
Comment (tomerXarchimedes) [inline test/VaultOUSD.js:40]: I would recommend adding an `expect` t make sure addr1 got the shares

File: test/VaultOUSD.js [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+        await tokenOUSD.connect(addr2).approve(tokenVault.address, addr2Deposit);
+        await tokenVault.connect(addr2).deposit(addr2Deposit, sharesOwnerAddress)
```
Comment (tomerXarchimedes) [inline test/VaultOUSD.js:42]: I would recommend adding an `expect` t make sure addr2 got the shares

File: test/VaultOUSD.js [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+        it("Should return OUSD to be Vault's asset", async function () {
+            let vaultAsset = await tokenVault.asset()
+            expect(vaultAsset).to.equal(helper.addressOUSD)
+        })
+
+        it("Should have an updated total assets sum after deposit", async function () {
+            expect(await tokenVault.totalAssets()).to.equal(addr1Deposit + addr2Deposit)
+        })
+
+        it("Should have all shares under owners address", async function () {
```
Comment (tomerXarchimedes) [inline test/VaultOUSD.js:55]: nitpicking - `depositor` is probably more accurate than `owners` (Owner has some other meaning around access control)

File: test/VaultOUSD.js [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+            expect(await tokenVault.maxRedeem(sharesOwnerAddress)).to.equal(addr1Deposit + addr2Deposit)
+        })
+
+        describe("Adding more money to vault as interest (ie no shares are minted)", function () {
+            before(async function () {
+                //increase Vaults balance without minting more shares
+                await tokenOUSD.transfer(tokenVault.address, interestIntoVault)
```
Comment (tomerXarchimedes) [inline test/VaultOUSD.js:62]: I would add an `expect` to ensure transfer was successful (like: balance check or at least check that vault balance increased)

File: test/VaultOUSD.js [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+                expect(await tokenVault.totalAssets()).to.equal(addr1Deposit + addr2Deposit + interestIntoVault)
+            })
+
+            it("Should not change number of shares per deposit", async function () {
+                /// Check the max number of share owner has 
+                expect(await tokenVault.maxRedeem(sharesOwnerAddress)).to.equal(addr1Deposit + addr2Deposit)
+            })
+
+            it("Should redeem with each share worth more then 1 underlying", async function () {
+                expect(await tokenVault.previewRedeem(addr1Deposit + addr2Deposit)).to.equal(addr1Deposit + addr2Deposit + interestIntoVault)
```
Comment (tomerXarchimedes) [inline test/VaultOUSD.js:75]: looks like we expect `previewRedeem` to calculate how much OUSD add1 and addr2 have to their balance. There is no explicit "shares" calculation that we need to do (looks like `previewRedeem` is doing that under the hood).

Not sure if I got it right - but I would suggest adding a comment explaining that

File: test/zIntegrationTests.js [hunk 8]
```diff
@@ -15,6 +15,11 @@ describe("Setting the stage: Getting some OUSD and deploying our contracts", fun
 
     let signer;
     let user;
+    
+    before(async function () {
+        /// Reset network before integration tests 
+        helper.helperResetNetwork(14533286)
```
Comment (tomerXarchimedes) [inline test/zIntegrationTests.js:21]: Nice! great catch :-)

**Review by tomerXarchimedes — CHANGES_REQUESTED** [review]
Good stuff - add some comments, nothing significant, but requesting changes just to keep it clean

File: test/VaultOUSD.js (hunk 2 above)
Comment (yotamdaniel) [reply test/VaultOUSD.js:25]: I am talking about 10k OUSD. Is the expectation for 10k OUSD to be returned and transfer via (10k*10^18)?

File: test/VaultOUSD.js (hunk 4 above)
Comment (yotamdaniel) [reply test/VaultOUSD.js:42]: Shares actually go to sharesOwnerAdress which is owner.address.

I am checking that shares got to owner.address on line 55 (  it("Should have all shares under owners address")

File: test/VaultOUSD.js (hunk 5 above)
Comment (yotamdaniel) [reply test/VaultOUSD.js:55]: see comment above, it not the depositors that get the shares

File: test/VaultOUSD.js (hunk 6 above)
Comment (yotamdaniel) [reply test/VaultOUSD.js:62]: added

File: test/VaultOUSD.js (hunk 7 above)
Comment (yotamdaniel) [reply test/VaultOUSD.js:75]: Thats right. Adding a comment

File: test/VaultOUSD.js (hunk 2 above)
Comment (tomerXarchimedes) [reply test/VaultOUSD.js:22]: OUSD is using 10^18 decimal
so the `transfer` function expects 10K*10^18 - if you keep it 10,000 just add a comment. It might be better to always use X*10^18 so we won't need to convert back and forth

**Review by tomerXarchimedes — COMMENTED** [review]
Added a comment around decimal usage

TLDR - we might want to agree to explict way to deal with decimals. Like: always using X*10^18 notation.

I am worried that we might forget to do the conversion and introduce a significant bug, so I believe it should be explicit

File: test/zIntegrationTests.js [hunk 9]
```diff
@@ -16,17 +16,22 @@ describe("Setting the stage: Getting some OUSD and deploying our contracts", fun
...
         [signer, user, addr2, ...addrs] = await ethers.getSigners();
 
     });
 
     it("Should do a basic ETH<>OUSD swap", async function () {
-        
-       await helper.helperSwapETHWithOUSD(user, ethers.utils.parseEther("100.0"))
-        
+
+        await helper.helperSwapETHWithOUSD(user, ethers.utils.parseEther("1.0"))
```
Comment (yotamdaniel) [inline test/zIntegrationTests.js]: Changing this since OUSD contract allows up to 25000  USD swap

**Comment by yotamdaniel** [comment]
Addressed comments about using 18 decimals (everything is in big 18 decimal numbers now)

**Review by tomerXarchimedes — APPROVED** [review]
Great job - thanks!
